scalding-core/src/main/scala/com/twitter/scalding/typed/HashJoinable.scala

```diff
@@ -50,11 +51,24 @@ trait HashJoinable[K, +V] extends CoGroupable[K, V] with KeyedPipe[K] {
       val newPipe = new HashJoin(
         RichPipe.assignName(mapside.toPipe(('key, 'value))(fd, mode, tup2Setter)),
         Field.singleOrdered("key")(keyOrdering),
-        mapped.toPipe(('key1, 'value1))(fd, mode, tup2Setter),
+        getMappedPipe(fd, mode),
         Field.singleOrdered("key1")(keyOrdering),
         WrappedJoiner(new HashJoiner(joinFunction, joiner)))
 
       //Construct the new TypedPipe
       TypedPipe.from[(K, R)](newPipe.project('key, 'value), ('key, 'value))(fd, mode, tuple2Converter)
     })
+
+  /**
+   * Returns a Pipe for the mapped (rhs) pipe. If this hasn't been persisted to disk,
+   * we force a checkpoint. Currently this persistence check is trivial: if the pipe is
+   * a ForceToDiskTypedPipe we skip an additional checkpoint.
+   */
+  def getMappedPipe[R, V1](fd: FlowDef, mode: Mode): Pipe = {
+    val mappedPipe = mapped.toPipe(('key1, 'value1))(fd, mode, tup2Setter)
+    mapped match {
```

Comment:
the better solution looks something like:
```scala
def forceBeforeFlatMap[T](tp: TypedPipe[T]): TypedPipe[T] = tp match {
  case f@ForceToDiskTypedPipe(_) => f // can't move something the user did
  case tpi : TypedPipeInst[_] =>
    // need to clean up this API to avoid the cast:
   val tpit = tpi.asInstanceOf[TypedPipeInst[T]]
   tpit.flatMapFn match {
     case FlatMappedFn(prev, f) =>
       val upstream = new TypePipeInst(pipe, ... , prev)
       val forced = forceBeforeFlatMap(upstream).
       // apply the flatMap
       forced.flatMap(f)
    case Converter(_) => tpit // this is just a source, already forced
    case Empty => TypedPipe.empty // may never happen, but for completeness
    case _ => tpit.forceToDisk
   }
   case p => p.forceToDisk
}
```
The goal here is to not materialize a big flatMap on disk. We do want to materialize after a filter or even a map (which is often a projection). We don't want a forceToDisk on a source, that is a waste.

to be more correct, until we have a better AST, at the point of calling `toPipe` we should check if that pipe is an instance of:

https://github.com/Cascading/cascading/blob/2.7/cascading-core/src/main/java/cascading/pipe/GroupBy.java#L74

or:

https://github.com/Cascading/cascading/blob/2.7/cascading-core/src/main/java/cascading/pipe/CoGroup.java#L114

in which case we don't need to checkpoint.

See https://github.com/twitter/scalding/blob/develop/scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala#L701

you can call `getPrevious` and if you don't hit any `Each` before hitting an `Every` you are okay.

To answer your question about going iteratively, I am nervous about that. This is a change that could hurt a lot of people (and without exaggeration cost millions of dollars) if done poorly. Let's take a little time and do it right.

I can help.